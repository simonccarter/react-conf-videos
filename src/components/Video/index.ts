import { Video } from './Video'

export {
  Video
}
