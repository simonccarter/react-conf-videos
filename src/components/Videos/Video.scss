@import '../../styles/_colors';

.root {
  width: 100%;
  padding: 1rem 2rem;
  box-sizing: border-box;
  font-size: 1.6rem;

  .top {
    display: flex;
    justify-content: space-between;
    cursor: pointer;

    .title {
      font-weight: bold;
    }

    .right {
      font-size: 1.6rem;
    }
  }

  .videoWrapper {
    height: 0;
    width: 100%;
    overflow-y: hidden;
    transition: height .5s;
    background: url('../../../public/assets/tv.svg');
    background-repeat: no-repeat;
    background-size: 150px 150px;
    background-position: center center;

    &.open {
      height: 360px;
      margin: 1rem 0;
      background-color: black;
    }
  }

  .details {
    display: flex;
    justify-content: space-between;
    font-size: 1.4rem;

    span {
      font-size: 1.4rem;
    }    
  }

  .divider {
    height: 1px;
    width: 100%;
    background-color: grey;
  }
}