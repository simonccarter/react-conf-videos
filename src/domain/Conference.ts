export type Conference = {
  date: string
  title: string
  website: string
  playlist: string
  videos: string[]
}

export type IndexedConferences = {
  [idx: string]: Conference
}

export type ConferenceTitlesToIds = {
  [idx: string]: string
}