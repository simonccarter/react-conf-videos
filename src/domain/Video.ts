export type Video = {
  link: string
  title: string
  split: string
  length: string
  presenter: string // id
  embeddableLink: string
}

