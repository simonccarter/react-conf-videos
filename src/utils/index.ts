export * from './test'
export * from './search'
export * from './url'