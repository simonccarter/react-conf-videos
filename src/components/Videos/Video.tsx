import * as React from 'react'
import * as cn from 'classnames'
import { connect } from 'react-redux'
import { compose, pure, withStateHandlers } from 'recompose'

import { Video as VideoProp, Presenter, Conference } from '../../domain'

import styles from './Video.scss'

type Props = {
  videoId: string
  conferenceId: string
}

type RecState = {
  isOpen: boolean
}

type StateHandlers = {
  toggleIsOpen: (e: any) => RecState
}

type RedState = {
  video: VideoProp
  conference: Conference
  speaker: Presenter
}

type CombinedProps = Props & RecState & RedState & StateHandlers

const VideoInner: React.SFC<CombinedProps> = ({
  video, speaker, videoId, conference, isOpen, toggleIsOpen
}) => {
  const {
    title, length, link, embeddableLink
  } = video
  return (
    <div className={styles.root} key={videoId} >
      <div className={styles.top} onClick={toggleIsOpen} >
        <a className={styles.title} href={link}>{title}</a>
        <span className={styles.right}>{length}</span>
      </div>
      <div className={cn(styles.videoWrapper, { [styles.open]: isOpen })}>
        {isOpen && <iframe
          title="videoPlayer"
          id="ytplayer"
          width="410"
          height="360"
          src={embeddableLink}
          frameBorder="0"
          allowFullScreen
        /> }
      </div>
      <div className={styles.details}>
        <span>{speaker.name}</span>
        <span>{conference.title}</span>
      </div>
    </div>
  )
}

const mapStateToProps = (state: any, props: Props) => {
  const { data: { videos, presenters, conferences } } = state
  const { videoId, conferenceId } = props
  const video = videos[videoId]
  const conference = conferences[conferenceId]
  const speaker = presenters[video.presenter]
  return {
    video: videos[videoId],
    conference,
    speaker
  }
}

const Video = compose<CombinedProps, Props>(
  connect(mapStateToProps, null),
  pure,
  withStateHandlers({ isOpen: false }, {
    toggleIsOpen: ({ isOpen }) => () => ({ isOpen: !isOpen })
  })
)(VideoInner)

export { Video }
