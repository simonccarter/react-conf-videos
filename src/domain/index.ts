import { Action } from './Action'
import { Video, IndexedVideos } from './Video'
import { Presenter, IndexedPresenters } from './Presenter'
import { Conference, IndexedConferences, ConferenceTitlesToIds } from './Conference'
import { ConferenceInput, VideoInput, PresenterInput, JSONInput } from './InputJSON'

export {
  Action,
  Video, IndexedVideos,
  Presenter, IndexedPresenters,
  Conference, IndexedConferences, ConferenceTitlesToIds,
  ConferenceInput, VideoInput, PresenterInput, JSONInput
}