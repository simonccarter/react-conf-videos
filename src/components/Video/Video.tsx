import * as React from 'react'
import * as cn from 'classnames'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { compose, pure, withHandlers, withStateHandlers } from 'recompose'

import { Video as VideoProp, Presenter, Conference } from '../../domain'
import { conferencePageActions } from 'redux/modules'
import { sluggifyUrl } from 'utils'

import styles from './Video.scss'

type Props = {
  videoId: string
  conferenceId: string
}

type State = {
  isOpen: boolean
}

type StateHandlers = {
  toggleIsOpen: (e: any) => State
}

type WithHandlers = { onConferenceClick: React.EventHandler<any> }

type ReduxState = {
  video: VideoProp
  conferenceId: string
  conference: Conference
  speaker: Presenter
  goToConfPage: typeof conferencePageActions.navigateToConferencePage
}

type CombinedProps = Props & State & ReduxState & StateHandlers & WithHandlers

export const VideoInner: React.SFC<CombinedProps> = ({
  video, speaker, videoId, conference, isOpen, toggleIsOpen, onConferenceClick, conferenceId
}) => {
  const {
    title, length, link, embeddableLink
  } = video
  return (
    <div className={styles.root} key={videoId} >
      <div className={styles.top} onClick={toggleIsOpen} >
        <a className={styles.title} href={link}>{title}</a>
        <span className={styles.right}>{length}</span>
      </div>
      <div className={cn(styles.videoWrapper, { [styles.open]: isOpen })}>
        {isOpen && <iframe
          title="videoPlayer"
          id="ytplayer"
          width="410"
          height="360"
          src={embeddableLink}
          frameBorder="0"
          allowFullScreen
        /> }
      </div>
      <div className={styles.details}>
        <span>{speaker.name}</span>
        <Link to={`/conference/${conferenceId}/${sluggifyUrl(conference.title)}`} onClick={onConferenceClick}>
          <span className={styles.conferenceTitle}>{conference.title}</span>
        </Link>
      </div>
    </div>
  )
}

const mapStateToProps = (state: any, props: Props) => {
  const { data: { videos, presenters, conferences } } = state
  const { videoId, conferenceId } = props
  const video = videos[videoId]
  const conference = conferences[conferenceId]
  const speaker = presenters[video.presenter]
  return {
    conferenceId,
    conference,
    speaker,
    video
  }
}

const mapDispatchToProps = {
  goToConfPage: conferencePageActions.navigateToConferencePage
}

const Video = compose<CombinedProps, Props>(
  connect(mapStateToProps, mapDispatchToProps),
  pure,
  withStateHandlers({ isOpen: false }, {
    toggleIsOpen: ({ isOpen }) => () => ({ isOpen: !isOpen })
  }),
  withHandlers<ReduxState & StateHandlers, WithHandlers>({
    onConferenceClick: props => (e: any) => {
      e.preventDefault()
      props.goToConfPage(props.conferenceId)
    }
  })
)(VideoInner)

export { Video }
