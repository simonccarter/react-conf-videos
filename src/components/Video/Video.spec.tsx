import * as React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'

import { shallowWithStore, mountWithStore } from 'utils'
import { Video, VideoInner } from './Video'
import { Video as VideoT, Conference } from '../../domain'

describe('Video', () => {
  describe('VideoInner', () => {
    const mockFN = jest.fn()

    const shallowVideo = () => {
      const video = {
        title: 'test title',
        length: '12:34',
        link: 'a link',
        embeddableLink: 'a link'
      }
      const conferenceId = 'conferenceId1234'
      const conference: Partial<Conference> = { title: 'A conference' }
      const videoId = 'videoId1234'
      const speaker = { name: 'Simon Carter' }
      const isOpen = false
      const toggleIsOpen = mockFN

      return shallow<any>(
        <VideoInner
          conferenceId={conferenceId}
          video={video as VideoT}
          videoId={videoId}
          speaker={speaker}
          isOpen={isOpen}
          toggleIsOpen={toggleIsOpen}
          conference={conference as Conference}
          goToConfPage={() => ({type: 'x', payload: 'x'})}
          onConferenceClick={() => null}
        />
      )
    }

    it('renders correctly', () => {
      // arrange
      const wrapper = shallowVideo()
      // act
      // assert
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('handles onClick', () => {
      // arrange
      const wrapper = shallowVideo()

      // act
      wrapper.find('.top').simulate('click')

      // assert
      expect(mockFN).toHaveBeenCalled()
    });
  })

  describe('Video', () => {

    const getData = () => {
      const props = { videoId: 'xxx', conferenceId: 'yyy'}
      const video = {
        title: 'test title',
        length: '12:34',
        link: 'a link',
        embeddableLink: 'a link',
        presenter: 'aaa'
      }
      const conference: Partial<Conference> = { title: 'A conference' }
      const speaker = { name: 'Simon Carter' }
      const state = {data:{
        videos: {'xxx': video},
        presenters: {'aaa': speaker},
        conferences: {'yyy' : conference}
      }}
      return { props, state, video, speaker, conference }
    }

    it('should render and connect', () => {
      // arrange
      const { props, state, video, speaker, conference } = getData()

      // act
      const wrapper = shallowWithStore(state, <Video {...props} />)

      // assert
      expect(toJSON(wrapper)).toMatchSnapshot()
      expect(wrapper.props().video).toEqual(video)
      expect(wrapper.props().conference).toEqual(conference)
      expect(wrapper.props().speaker).toEqual(speaker)
    })

    it('should toggle the isOpen state prop on click', () => {
      // arrange
      const { props, state } = getData()

      // act
      const wrapper = mountWithStore(state, <Video {...props} />)
      wrapper.find('.top').simulate('click')

      // assert
      expect(wrapper.find('.root').parent().props().isOpen).toEqual(true)
    })

  })
})


