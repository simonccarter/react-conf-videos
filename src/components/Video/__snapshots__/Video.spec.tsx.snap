// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video Video should render and connect 1`] = `
<pure(withStateHandlers(Component))
  conference={
    Object {
      "title": "A conference",
    }
  }
  conferenceId="yyy"
  dispatch={[Function]}
  speaker={
    Object {
      "name": "Simon Carter",
    }
  }
  video={
    Object {
      "embeddableLink": "a link",
      "length": "12:34",
      "link": "a link",
      "presenter": "aaa",
      "title": "test title",
    }
  }
  videoId="xxx"
/>
`;

exports[`Video VideoInner renders correctly 1`] = `
<div
  className="root"
  key="videoId1234"
>
  <div
    className="top"
    onClick={[MockFunction]}
  >
    <a
      className="title"
      href="a link"
    >
      test title
    </a>
    <span
      className="right"
    >
      12:34
    </span>
  </div>
  <div
    className="videoWrapper"
  />
  <div
    className="details"
  >
    <span>
      Simon Carter
    </span>
    <span>
      A conference
    </span>
  </div>
</div>
`;
